File: 최소 신장 트리/4386. 별자리 만들기.py
import sys
import math

def getParent(parents, i):
    if parents[i] == i:
        return i
    return getParent(parents, parents[i])

def unionParent(parents, a, b):
    parentA = getParent(parents, a)
    parentB = getParent(parents, b)
    if parentA > parentB: # B가 진짜 부모
        parents[a] = b
    else:
        parents[b] = a

def main():
    n = int(input())
    stars = []
    edges = []
    parents = [i for i in range(n)]

    for _ in range(n):
        x, y = map(float, sys.stdin.readline().split())
        stars.append((x, y))

    for i in range(n-1):
        for j in range(i+1, n):
            x = stars[i][0] - stars[j][0]
            y = stars[i][1] - stars[j][1]
            dist = math.sqrt(x**2 + y**2)
            edges.append((dist, i, j))

    edges.sort() # dist에 따라 오름차순으로 정렬
    print(edges)

    ans = 0
    for dist, i, j in edges:
        print(parents)
        if getParent(parents, i) != getParent(parents, j):
            print(parents)
            unionParent(parents, i, j)
            ans += dist

    print(dist)

if __name__ == "__main__":
    main()

File: BaekJoon/최소 신장 트리/4386_별자리 만들기.swift
// 별의 개수 100개
// 별들을 잇는 간선의 개수 100^2 = 10000개
import Foundation

let n = Int(readLine()!)!
var stars = [[Float]]()
var roots = [Int](repeating: 0, count: n + 1)
var dists: [(Float, Int, Int)] = []

func getParent(_ current: Int) -> Int {
    if roots[current] != current {
        roots[current] = getParent(roots[current])
    }
    
    return roots[current]
}

func unionParent(_ left: Int, _ right: Int) {
    let p1 = getParent(left)
    let p2 = getParent(right)
    
    if p1 < p2 {
        roots[p2] = p1
    } else {
        roots[p1] = p2
    }
}

func findParent(_ left: Int, _ right: Int) -> Bool {
    let p1 = getParent(left)
    let p2 = getParent(right)
    
    return p1 == p2
}

// 두 별 사이의 거리를 계산하는 메서드
func getDistance(_ x1: Float, _ y1: Float, _ x2: Float, _ y2: Float) -> Float {
    return sqrt(pow((x2 - x1), 2) + pow((y2 - y1), 2))
}

for i in 1...n {
    roots[i] = i
}

for _ in 0..<n {
    stars.append(readLine()!.split(separator: " ").map { Float(String($0))! })
}

// 모든 별 사이의 거리를 계산
for i in 1..<n {
    for j in i + 1...n {
        let dist = getDistance(stars[i - 1][0], stars[i - 1][1], stars[j - 1][0], stars[j - 1][1])
        dists.append((dist, i, j))
    }
}

dists.sort(by: { $0.0 > $1.0 }) // removeLast를 위해 내림차순으로 정렬

var ans: Float = 0

while !dists.isEmpty {
    let current = dists.removeLast()
    
    if !findParent(current.1, current.2) {
        unionParent(current.1, current.2)
        ans += current.0
    }
}

print(round(ans * 100) / 100)
